### PR TITLE
Add IAsyncEnumerable<T> support

### LINKED ISSUE #328: Add support for IAsyncEnumerable<T>
[`IAsyncEnumerable<T>` is here](https://apisof.net/catalog/System.Collections.Generic.IAsyncEnumerable%3CT%3E) and it's a more natural fit than `Progress<T>` when the RPC method wants to exchange data *as* its result instead of simply reporting progress *about* its result.

We should add support for it.

For downlevel framework versions, this type is defined in the [Microsoft.Bcl.AsyncInterfaces](https://www.nuget.org/packages/Microsoft.Bcl.AsyncInterfaces) package.

In the spec here, we use the term `generator` to refer to the app-level implementation of `IAsyncEnumerable<T>` that produces the values, and `consumer` to refer to the app-level enumeration of those values. We avoid the terms `server` and `client` because the `IAsyncEnumerable<T>` may be passed to the server via an RPC method argument or passed to the client via an RPC return value.

## Sample use case

A server method which returns an async enumeration may be defined as:

```cs
IAsyncEnumerable<int> GenerateNumbersAsync(CancellationToken cancellationToken);
```

Per C# 8 syntax this may be consumed via:

```cs
await foreach (int number in this.clientProxy.GenerateNumbersAsync(token))
{
    Console.WriteLine(number);
}
```

The server method may be implemented in C# 8 as:

```cs
public async IAsyncEnumerable<int> GenerateNumbersAsync([EnumeratorCancellation] CancellationToken cancellationToken)
{
    for (int i = 1; i <= 20; i++)
    {
        cancellationToken.ThrowIfCancellationRequested();
        await Task.Yield();
        yield return i;
    }
}
```

We must support methods which return `IAsyncEnumerable<T>`.
We *may* also support methods which take `IAsyncEnumerable<T>` as parameters.

## State machine internals

The following describes how `IAsyncEnumerable<T>` return types will be supported. `IAsyncEnumerable<T>` method parameters may be supported later.

1. Client invocation results in an RPC call, and the server invokes the method that returns an `IAsyncEnumerable<T>`. A C# async-iterator method will return this object without executing user code within the method. The resulting `IAsyncEnumerable<T>` object will be immediately converted to an `IAsyncEnumerator<T>` object by the RPC server and this object will be retained. A handle to this enumerator instance will be returned to the RPC client.
1. The RPC client exposes an `IAsyncEnumerable<T>` object to the caller, whose `GetAsyncEnumerator` method will only work once. Subsequent calls to this method will throw `InvalidOperationException`. To repeat the enumeration, the RPC client should call the RPC method again.
1. When the RPC client application calls `IAsyncEnumerable<T>.GetAsyncEnumerator(CancellationToken)`, the RPC client library produces an `IAsyncEnumerator<T>` instance which handles the RPC layer and emulates a local async enumerator. The `CancellationToken` passed to it applies to each RPC call to the server when asking for more value(s).
1. When the client application calls `DisposeAsync` on the `IAsyncEnumerator<T>`, an RPC call is made to the server so the server can release the memory holding the server's original `IAsyncEnumerator<T>` in memory.

If the originating request or a subsequent request for values is ever canceled, this cancels the entire enumerator. It is not possible to obtain more values after a request for more values is canceled. However clients must still dispose of the enumerator if they ever received the enumerator handle in response to their originating request.

When the client makes the request on a dynamically generated proxy, and the method returns `IAsyncEnumerable<T>` directly (not wrapped in `Task<T>`), the RPC client proxy will have to create an `IAsyncEnumerable<T>` implementation that is available immediately, and ends up waiting for the async RPC call to complete before returning from the proxy's private implementation of `IAsyncEnumerator<T>.MoveNextAsync()`.

## Performance tuning

The `IAsyncEnumerator<T>` interface is defined assuming that every single value produced may be acquired asynchronously. Thus an RPC implementation of this feature might be that every request for a value results in an RPC call to request that value. Such a design would minimize memory consumption and cost of generating values unnecessarily, but would be particularly noisy on the network and performance would suffer on high latency connections. 

To improve performance across a network, the RPC layer will support:

1. Batching: When the consumer requests a value from the generator, the server may respond with the next several values in the same response message in order to reduce the number of round-trips the client must make while enumerating the sequence. This improves performance when network latency is significant.
1. Read ahead: The generator will do work to produce the next value(s) _before_ receiving the consumer's request for them. This allows for the possibility that the server is processing the data while the last value(s) are in transit to the client or being processed by the client. This improves performance when the time to generate the values is significant.

The above optimizations are configured individually, and may be used in combination.

To accomplish this, we define these properties which tune the RPC bridge for async enumerables:

```cs
        /// <summary>
        /// Gets or sets the maximum number of elements to read ahead and cache from the generator in anticipation of the consumer requesting those values.
        /// </summary>
        public int MaxReadAhead { get; set; }

        /// <summary>
        /// Gets or sets the minimum number of elements to obtain from the generator before sending a batch of values to the consumer.
        /// </summary>
        public int MinBatchSize { get; set; } = 1;
```

The default values for these properties will result in the same behavior as one would observe with `IAsyncEnumerable<T>` without any RPC: one value is produced or consumed at a time. 
Individual services may choose based on expected use cases and performance costs to change these values to improve the user experience.

These properties are defined on the new `JsonRpcEnumerableAttribute`, which may be applied on the  return type of a server method implementation. It will be ignored if used within an interface declaration. For example, the aforementioned C# 8 server method may be defined like this in order to provide all the values to the client in batches of 10:

```cs
[return: JsonRpcEnumerable(MinBatchSize = 10)]
public async IAsyncEnumerable<int> GenerateNumbersAsync([EnumeratorCancellation] CancellationToken cancellationToken)
{
    for (int i = 1; i <= 20; i++)
    {
        cancellationToken.ThrowIfCancellationRequested();
        await Task.Yield();
        yield return i;
    }
}
```

A batch size of 10 with the default `MaxReadAhead` value of `0` means the server will not produce any values until the client requests them, but when the client requests them, the client will get 10 or the rest of the sequence, whichever is fewer.

Suppose `MaxReadAhead = 15` and `MinBatchSize = 10`. After the client calls `GenerateNumbersAsync` but before it asks for the first element in the sequence, the server is already generating values till it fills a cache of 15. When the client request comes in for the first value(s), the server will wait till at least 10 items have been produced before returning any to the client. If more than 10 were cached (e.g. the server was able to produce all 15 as part of "read ahead"), the client will get all that are cached. After the client's request is fulfilled, the read ahead server will continue generating items till the read ahead cache is full, in preparation for the next client request.

In short: `MinBatchSize` guarantees a minimum number of values the server will send to the client except where the sequence is finished, and `MaxReadAhead` is how many values may be produced on the server in anticipation of the next request from the client for more values.

The state machine and any cached values are released from the generator when the `IAsyncEnumerator<T>` is disposed.

### Originating request 

The original RPC request to the `IAsyncEnumerable<T>` returning method includes a special `expect: "enumerableResponse"` property to indicate to the server that it is capable of understanding the lifetime requirements of the returned object. This is important because upon successful completion of this particular RPC call, the server will now retain memory on behalf of the client that may be a memory leak if the client never follows up with another call to the server to release the memory. 
An RPC call to an `IAsyncEnumerable<T>`-returning method that lacks this property should result in the server returning an error response and not invoking the method.

```json
{
   "jsonrpc": "2.0",
   "id": 1,
   "method": "GenerateNumbersAsync",
   "params": [ "some", "parameters"],
   "expect": "enumeratorResponse",
}
```

If the request omits the `id` property, the client will not get a result, and the server should either reject the request or execute the method but discard the result without leaking memory.

### Originating result

A successful result provides the handle to the server's enumerator, which may be any valid JSON token except `null`:

```json
{
   "jsonrpc": "2.0",
   "id": 1,
   "result": "enum-handle"
}
```

### Client request for values

A request for value(s) from the client:

```json
{
   "jsonrpc": "2.0",
   "id": 2,
   "method": "rpc.Enumerator.next",
   "params": { "token": "enum-handle" },
}
```

### Server response with values

A response with value(s) from the server always comes in an array, even if only one element is returned:

```json
{
   "jsonrpc": "2.0",
   "id": 2,
   "result": {
      "values": [ 1, 2, 3 ],
      "finished": false
   }
}
```

Note the `finished` property is a hint from the server for when the last value has been returned. 
The server *may* not know that the last value has been returned and should specify `false` unless it is sure the last value has been produced. 

The client *may* ask for more values without regard to the `finished` property, but if `finished: true` the client may optimize to not ask for more values but instead go directly to dispose of the enumerator.

The server should never return an empty array of values unless the last value in the sequence has already been returned to the client.

### Client disposes enumerator

The client always notifies the server when the local enumerator proxy is disposed. This is typically a notification (it has no `id` property).

```json
{
   "jsonrpc": "2.0",
   "method": "rpc.Enumerator.dispose",
   "params": { "token": "enum-handle" },
}
```

### PR DESCRIPTION
**Pull request by AArnott**
Closes #328

### CODE REVIEW

**Comment by jepetty** [comment]
I appreciate your previous PRs! I see a lot of similarities between this addition and `IProgress<T>`, so made it easier to review than the size of changes might suggest. :)

File: src/StreamJsonRpc/Reflection/MessageFormatterEnumerableTracker.cs [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+
+            ValueTask<object> GetNextValuesAsync(CancellationToken cancellationToken);
+        }
+
+        /// <summary>
+        /// Checks if a given <see cref="Type"/> implements <see cref="IAsyncEnumerable{T}"/>.
+        /// </summary>
+        /// <param name="objectType">The type which may implement <see cref="IAsyncEnumerable{T}"/>.</param>
+        /// <returns>true if given <see cref="Type"/> implements <see cref="IAsyncEnumerable{T}"/>; otherwise, false.</returns>
+        public static bool CanSerialize(Type objectType) => TrackerHelpers<IAsyncEnumerable<int>>.CanSerialize(objectType);
```
[the remark this answers is not in the document]
Comment (AArnott) [reply src/StreamJsonRpc/Reflection/MessageFormatterEnumerableTracker.cs]: Done.

File: src/StreamJsonRpc/Reflection/MessageFormatterEnumerableTracker.cs [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+
+        /// <summary>
+        /// Handles the <see cref="IJsonRpcFormatterCallbacks.ErrorReceived"/> event.
+        /// </summary>
+        /// <param name="sender">The <see cref="JsonRpc"/> instance that raised the event.</param>
+        /// <param name="e">The event args.</param>
+        private void JsonRpc_ErrorReceived(object sender, JsonRpcFormatterCallbackEventArgs e) => this.CleanUpResources(e.RequestId, error: true);
+
+        /// <summary>
+        /// Handles the <see cref="IJsonRpcFormatterCallbacks.ErrorReceived"/> event.
```
[the remark this answers is not in the document]
Comment (AArnott) [reply src/StreamJsonRpc/Reflection/MessageFormatterEnumerableTracker.cs]: good eyes

**Comment by AArnott** [comment]
Thanks so much. Yes, I asked you to review for just that reason. :)

**Pull request merged**